docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,354 @@
+  HVPA's policy for early horizontal scale-out.
+- Allow usage of VPA for right-sizing the resource requests during the entire time and not only under certain conditions.
+- Enable future migration away from HVPA by removing all dependency on it for scaling ShootKapi.
+- Reduce Gardener custom code base and ongoing maintenance effort.
+
+### Non-Goals
+There are other components currently scaled via HVPA. Changing the autoscaling mode of these, and of `kube-apiserver`s
+other than the one in the shoot control plane, is outside the scope of this GEP. These workloads are subject to a
+different mode of autoscaling (e.g. vertical-only, where HVPA is only used to stabilize VPA behavior), and it is
+advantageous from project execution perspective to address them as a separate concern.
+
+## Proposal
+ShootKapi HVPA is replaced by individual HPA and VPA. Undesirable interference between HPA and VPA is avoided through
+the use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is
+driven by CPU and memory usage metrics.
+
+### Rationale
+A rough, heuristic horizontal sizing sets the stage for optimal VPA operation (horizontal size is within stable and
+reasonably efficient range). Vertical scaling fine-tunes requests, ensuring high efficiency (requests not too high)
+and stability (requests not too low).
+
+### Design Outline
+One HPA and one VPA resources are deployed alongside each ShootKapi deployment object. The VPA controls CPU and memory
+requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
+deployment.
+
+A new component named gardener-custom-metrics is deployed on each seed. It directly scrapes metrics data from all
+ShootKapi pods on the seed and derives custom metrics based on it. The component is registered as an extension API
+service to the seed kube-apiserver. It occupies the custom metrics API extension point and is responsible for providing
+all custom metrics for the seed kube-apiserver, including the one driving the ShootKapi HPA.
+
+![Design outline](./assets/gep22-design-outline.png "Design outline")
+
+_Fig 2: Proposed design_
+
+### Element: Gardener Custom Metrics Provider Component 
+A new component, named gardener-custom-metrics is added to seed clusters. It periodically scrapes the metrics endpoints
+of all ShootKapi pods on the seed. It implements the K8s custom metrics API and provides K8s metrics specific to
+Gardener, based on custom calculations. The proposed design can naturally be extended to multiple sources of
+input data and a non-volatile cache for acquired data. However, in the scope of this proposal, the only data source
+is the aforementioned metrics scrape, and the calculated values only need to be briefly cached in memory.
+
+A single replica is deployed, as highly available autoscaling setup is currently prevented by limitations in
+the autoscalers utilized by Gardener.
```

Comment:
What does this mean? Which limitations?

Reply:
VPA doesn't have HA support. Multiple instances can work only if they are configured to target different namespaces. Running multiple replicas of the gardener-custom-metrics would still result in a minuscule improvement of expected availability, but it's not worth the extra running cost, since it wouldn't suffice for a full HA setup, with VPA's single replica.

I could be more specific. Is the following an improvement:
_A single replica of the gardener-custom-metrics is deployed. HA operation is not currently supported for VPA. That prevents an overall HA autoscaling setup and renders the use of multiple gardener-custom-metrics replicas economically unjustified._

Reply:
I don't get this discussion.
We can deploy 2 replicas + PDB and it would work with VPA. I don't see an issue.

Reply:
VPA itself does not support leader election, so it's not possible to run it in K8s' typical active/passive HA arrangement. Since we don't get highly available VPA, we don't get nominally HA autoscaling. So it's not justified to pay the cost of extra gardener-custom-metrics replicas, when that would still not result in nominally HA autoscaling.

I'm saying "nominally", because autoscaling generally does not require fast reactions. The time required to spin up a new VPA controller pod is an order of magnitude less than delays introduced by rate limiting on a loaded cluster.

The only scenario in which I know that lack of VPA controller redundancy could have impact is:
- VPA controller is in AZ1
- AZ1 fails
- There is lasting capacity shortage in AZ2, so the VPA controller cannot be scheduled

Reply:
I still don't get it. 
If VPA can't use leader-election so that we need to deploy a single VPA replica, why can't we still have the custom-metrics-server (which happens to be scaled by VPA) deployed in HA mode (two replicas)?

Reply:
We could have multiple replicas of gardener-custom-metrics. It's just not worth the expense (IMO), since it does not achieve nominally HA autoscaling functionality. Gardener-custom-metrics is very similar to VPA - it starts quickly and its absence does not immediately disrupt availability. So it's hard to justify paying for redundant replicas if you'd still not be able to declare HA.

Reply:
Nothing stops gardener-custom-metrics from running in a most reliable active/active replication mode, but cost. The extra replica(s) would have minimal contribution to risk reduction, but would impart substantial cross-AZ traffic.